(v9) Add JWT auth guide for ElasticSearch (#12612)
gravitational/teleport: docs/pages/application-access/jwt/elasticsearch.mdx

  ElasticSearch supports JWT authentication starting from version 8.2.0 and
  as of this writing is in beta.
</Details>

## Prerequisites

- Teleport cluster version >= 9.3 with configured [Application Access](../guides/connecting-apps.mdx).
- ElasticSearch cluster.

## Step 1. Enable JWT realm in ElasticSearch

Update your ElasticSearch configuration file `elasticsearch.yaml` to enable a
JWT realm:

```yaml
xpack.security.authc.realms.jwt.jwt1:
  order: 1
  client_authentication.type: none
  pkc_jwkset_path: https://proxy.example.com/.well-known/jwks.json
  claims.principal: sub
  claims.groups: roles
  allowed_issuer: example-cluster
  allowed_audiences: ["https://elasticsearch.example.com:9200"]
```

Let's take a closer look at the parameters and their values:

- Set `client_authentication.type` to `none`, otherwise ElasticSearch requires
  clients to send a shared secret value with each request.
- Set `pkc_jwkset_path` to the JWT key set file URL of your Teleport Proxy.
  You can also download the JSON file from the same URL and point the path
  directly to it instead of using a URL.
- Set `claims.principal` and `claims.groups` to `sub` and `roles` respectively.

>   As described in [Introduction](./introduction.mdx), these are the claims

Reviewer: ```suggestion
These are the claims
```
We should link to `introduction.mdx` in "Next steps" rather than here so the user doesn't need to navigate away from this page.

  Teleport uses to pass user and roles information in JWT tokens. Keep in mind
  that **users and roles must exist** in ElasticSearch.
- Set `allowed_issuer` to the name of your Teleport cluster.
- Set `allowed_audiences` to the URL which Teleport Application Service will
  use to connect to ElasticSearch (from the application YAML configuration).

Note that when using JWT authentication, you cannot map user roles using
standard ElasticSearch `role_mapping.yml` file. Instead, you need to set the

> role mapping using the API. See [JWT realm authorization](https://www.elastic.co/guide/en/elasticsearch/reference/current/jwt-realm.html#jwt-authorization)

Reviewer: Will all users be editing Elasticsearch role mapping? If so, we might want to spell out all the relevant steps here. If not, I think we should put this paragraph in a `Details` or `Admonition` box to clarify that it has to do with a special case and isn't part of the main tutorial.

for details.

## Step 2. Register ElasticSearch application in Teleport

In your Teleport App Service configuration `teleport.yaml`, register an entry
for ElasticSearch:

```yaml
app_service:
  enabled: "yes"
  apps:
  - name: "elastic"
    uri: https://elasticsearch.example.com:9200
    rewrite:
      headers:
      - "Authorization: Bearer {{internal.jwt}}"
```

<Admonition type="tip">
  You can also use [dynamic registration](../guides/dynamic-registration.mdx).
</Admonition>

ElasticSearch requires JWT token to be passed inside `Authorization` header.
The header rewrite configuration above will replace `{{internal.jwt}}` template
variable with Teleport-signed JWT token in each request as described in

> [Inject JWT](./introduction.mdx#inject-jwt).

Reviewer: I think we can remove this link here since the rest of the paragraph is clear enough without it. We can then add the link to "Next steps."

> ## Step 3. Connect to ElasticSearch API

Reviewer: ```suggestion
## Step 3. Connect to the ElasticSearch API
```

> Log into your Teleport cluster with `tsh login` and make sure ElasticSearch

Reviewer: ```suggestion
Log into your Teleport cluster with `tsh login` and make sure your ElasticSearch
```

application is available:

```code
$ tsh app ls
Application Description   Public Address               Labels
----------- ------------- ---------------------------- -------------------------------
elastic                   elastic.teleport.example.com
```

> Fetch short-lived X.509 certificate for it:

Reviewer: ```suggestion
Fetch a short-lived X.509 certificate for Elasticsearch:
```

```code
$ tsh app login elastic
```

> Then you can use curl command to communicate with ElasticSearch API which

Reviewer: ```suggestion
Then you can use the `curl` command to communicate with the Elasticsearch API, which
```

---
title: Using JWT authentication with Elasticsearch
description: How to use JWT authentication with Elasticsearch
---

This guide will help you configure Elasticsearch [JWT authentication](https://www.elastic.co/guide/en/elasticsearch/reference/current/jwt-realm.html)
with Teleport Application Access.

<Details title="Version warning" opened={true}>
  Elasticsearch supports JWT authentication starting from version 8.2.0.
</Details>

## Prerequisites

> - Teleport cluster version >= 9.3 with running Auth/Proxy Services and configured [Application Access](../guides/connecting-apps.mdx).

Author: Yeah, in this case it is different enough that I don't think it's really worth it.

> - Teleport cluster version >= 9.3 with running Auth/Proxy Services and configured [Application Access](../guides/connecting-apps.mdx).

Reviewer: 👍 
I think we can clarify "configured Application Access." Does that mean that there's at least one host running the Application Service and at least one connected application?

- Elasticsearch cluster.

## Step 1. Enable a JWT realm in Elasticsearch

Update your Elasticsearch configuration file, `elasticsearch.yaml`, to enable a
JWT realm:

```yaml
xpack.security.authc.realms.jwt.jwt1:
  order: 1
  client_authentication.type: none
  pkc_jwkset_path: https://proxy.example.com/.well-known/jwks.json
  claims.principal: sub
  claims.groups: roles
  allowed_issuer: example-cluster

>   allowed_audiences: ["https://elasticsearch.example.com:9200"]

Author: In most cases yes. It probably can have more than 1 element if other applications also use JWT auth and connect to ElasticSearch using different address.

```

Let's take a closer look at the parameters and their values:

- Set `client_authentication.type` to `none`, otherwise Elasticsearch requires
  clients to send a shared secret value with each request.
- Set `pkc_jwkset_path` to the JWT key set file URL of your Teleport Proxy.
  You can also download the JSON file from the same URL and point the path
  directly to it instead of using a URL.
- Set `claims.principal` and `claims.groups` to `sub` and `roles` respectively.
  These are the claims Teleport uses to pass user and role information in JWT
  tokens. Keep in mind that **users and roles must exist** in Elasticsearch.
- Set `allowed_issuer` to the name of your Teleport cluster.
- Set `allowed_audiences` to the URL which Teleport Application Service will

>   use to connect to Elasticsearch (the `uri` field in the application YAML

Reviewer: Since we mention the `uri` field in the next section, should this specify the URL in a different way, e.g., "(the fully qualified domain name of your Elasticsearch cluster's HTTP API)"?

Author: Not sure I fully understood what you meant here. But rephrased "the URL" to "the address" in this sentence, maybe it's more clear this way.

Reviewer: Sorry about that—what I meant is that, if a reader is following the guide step by step, they won't know what "the uri field in the application YAML configuration" means until they get to Step 2, since that is where the guide first includes an example of the field.

I think we can remove the parenthesis here and explain any format restrictions we need to mention, e.g., if you need to include a URL scheme, FQDN, and port.

Author: Makes sense, I removed the mention of `uri` and brought back "the URL", I think it should be clear now.

---
title: Using JWT authentication with Elasticsearch
description: How to use JWT authentication with Elasticsearch
---

This guide will help you configure Elasticsearch [JWT authentication](https://www.elastic.co/guide/en/elasticsearch/reference/current/jwt-realm.html)
with Teleport Application Access.

<Details title="Version warning" opened={true}>
  Elasticsearch supports JWT authentication starting from version 8.2.0.
</Details>

## Prerequisites

- Teleport cluster version >= 9.3 with running Auth/Proxy Services and configured [Application Access](../guides/connecting-apps.mdx).
- Elasticsearch cluster.

> ## Step 1. Enable a JWT realm in Elasticsearch

Reviewer: ```suggestion
## Step 1/3. Enable a JWT realm in Elasticsearch
```
To be consistent with other guides

  clients to send a shared secret value with each request.
- Set `pkc_jwkset_path` to the JWT key set file URL of your Teleport Proxy. It
  is available at `https://<proxy>/.well-known/jwks.json` endpoint. You can
  also download the JSON file from the same URL and point the path directly to
  it instead of using a URL.
- Set `claims.principal` and `claims.groups` to `sub` and `roles` respectively.
  These are the claims Teleport uses to pass user and role information in JWT
  tokens. Keep in mind that **users and roles must exist** in Elasticsearch.
- Set `allowed_issuer` to the name of your Teleport cluster.
- Set `allowed_audiences` to the address which Teleport Application Service
  will use to connect to Elasticsearch (the `uri` field in the application YAML
  configuration).

<Admonition title="Elasticsearch role mapping" type="note">
  Note that when using JWT authentication, you cannot map user roles using the
  standard Elasticsearch `role_mapping.yml` file. Instead, you need to set the
  role mapping using the API. See [JWT realm authorization](https://www.elastic.co/guide/en/elasticsearch/reference/current/jwt-realm.html#jwt-authorization)
  for details.
</Admonition>

> ## Step 2. Register an Elasticsearch application in Teleport

Reviewer: ```suggestion
## Step 2/3. Register an Elasticsearch application in Teleport
```

In your Teleport App Service configuration file, `teleport.yaml`, register an
entry for Elasticsearch:

```yaml
app_service:
  enabled: "yes"
  apps:
  - name: "elastic"
    uri: https://elasticsearch.example.com:9200
    rewrite:
      headers:
      - "Authorization: Bearer {{internal.jwt}}"
```

<Admonition type="tip">
  You can also use [dynamic registration](../guides/dynamic-registration.mdx).
</Admonition>

Elasticsearch requires a JWT token to be passed inside the `Authorization`
header. The header rewrite configuration above will replace the `{{internal.jwt}}`
template variable with a Teleport-signed JWT token in each request.

> ## Step 3. Connect to the ElasticSearch API

Reviewer: ```suggestion
## Step 3/3. Connect to the ElasticSearch API
```
